src/librustc_trans/mir/block.rs

```diff
@@ -249,11 +249,11 @@ impl<'a, 'tcx> MirContext<'a, 'tcx> {
                     } else {
                         lvalue.llval
                     };
-                    MaybeSizedValue::sized(value)
+                    LvalueRef::new_sized_ty(value, ty)
                 } else {
-                    MaybeSizedValue::unsized_(lvalue.llval, lvalue.llextra)
+                    LvalueRef::new_unsized_ty(lvalue.llval, lvalue.llextra, ty)
```

Comment:
These two branches could reuse `lvalue` and only mutate the `llval` field if it needs a cast.